### DISCUSSION TITLE
Looking for Relations of Doors and Spaces

### DISCUSSION THREAD

**Discussion by DprOliv** [IfcOpenShell/IfcOpenShell#2713] [General]
Exploring a little on using of BoundedBy, RelatedOpeningElement, HasFillings, i created a loop (as below) with the intention to populate a dictionary called "SpacesAccess" with all doors which provides access to each space... 

<img width="773" alt="Screen Shot 2023-01-21 at 01 09 31" src="https://user-images.githubusercontent.com/122423170/213889856-8d1fb6d5-45a9-44ef-95e8-172dc442deca.png">
A quick note: I'm not sure if my approach above is efficient (any recommendations on how to improve that are very welcomed !) 

The loop is working (well partially...)  since it's finding the doors in each wall bounding the related spaces ... 
However... in this model there's a specific condition: a wall which divides the 3 rooms called "Sala 1", "Sala2" and "Sala 3"), and all those 3 rooms will provide mistakenly the 3 doors (P02, P03 and P04) as access for each of those rooms. Naturally this happened because of this single wall (highlighted in red below...) bounds each one of those rooms.

<img width="913" alt="Screen Shot 2023-01-20 at 23 34 42" src="https://user-images.githubusercontent.com/122423170/213890022-ffd741e4-a7b9-4db8-8ac5-1e547311502e.png">

Does anyone has an idea of any technique (i.g. some relation in IFC)  that could be used to avoid this situation (without having to split this wall in the original model to break it to bound separately to each space and re-export) ? Or the only possible approach to turn around this situation would be using BVH (ifcopenshell.geom.tree) ?

**Comment by aothms** [comment]
Are you sure there are no direct space boundary relationships from Space to Door?

**Comment by DprOliv** [comment]
Hi @aothms

No i'm not... ;-/  Oh well... that's awesome !  (follows below the new code ... way more direct and clean now...)... 
In fact, RelatedBuildingElement in space.BoundedBy will return me the Doors ...

But i confess that i'm kind of confused now... In which case would we need to use all the Opening / Voids/ Fills approach to get the Doors of Spaces ? As i was previously doing ?
 
<img width="536" alt="Screen Shot 2023-01-22 at 14 36 47" src="https://user-images.githubusercontent.com/122423170/213930984-cc0d7e9c-602f-4d16-bdf8-0f85c94944c1.png">


And thank you (once more) for your guidance !

**Reply by DprOliv** [reply]
Hi Thomas...
Just found out here that:
a-  RelatedBuildingElement in IfcRelSpaceBoundary was made mandatory only in IFC4 
b- The export of space boundaries is optional - it is not in Coordination View 2.0 MVD and it was out of the certification

So, in the situations above, usage of BVH is the only way to get this relations ... is this assumption correct ?
 
Thanks

**Reply by aothms** [reply]
Yes, but this applies to nearly everything in IFC. The more assumptions you make, the more they get violated. The same applies to the correctness of space boundaries, there tend to be issues as it's not one of the core aspects of the model that is immediately visible.

In future versions of IfcOpenShell (0.8 or 0.9) we'll likely have tools to automatically compute the space boundaries. That's how I think it should be. Good dedicated tools for analysis as opposed to every authoring tool having to implement this during export.

**Reply by DprOliv** [reply]
Yes, it makes a lot of sense Thomas,  having dedicated tools to generate the space boundaries here at IFCOpenShell, where we use it ... rather then leaving to the design tool to export it, as you just said. 

From here, i did some more experiences with IFC2x3 / IFC4 Vectorworks 2023
And in both schemas, all depends only in activating the "1st Level Space Boundary" option to export either walls and Doors to the IFCRelSpaceBoundary ... and the Doors are properly 

What was not clear to me was when to use the relations: Space BoundedBy > IFCWall > IFcOpening > IFCFills to reach the Doors related to the Walls  bounding the spaces... ?
I mean, if the relation of "Space Boundary (1st level) is activated"  the Doors are directly reached by BoundedBy (no mater if in 2x3 or 4 schema).

And if not selected (this option to export space boundary 1st Level) we may not use even Doors directly nor Walls > Opening > Filling > Doors
Only option is to make use of BVH will help 

Thank you

**Reply by aothms** [reply]
> What was not clear to me was when to use the relations: Space BoundedBy > IFCWall > IFcOpening > IFCFills to reach the Doors related to the Walls bounding the spaces... ?

I don't think this would be more reliable. I don't know what caused exploring this first. 

>  Only option is to make use of BVH will help

True. It's rather straightforward. An efficient (bvh with only spaces, windows as point queries) is provided here https://community.osarch.org/discussion/comment/8902/#Comment_8902

**Reply by DprOliv** [reply]
Thanks Thomas... 
I mentioned the approach using IfcOpening/Voids/Fills in Walls bounding spaces because this is pretty widely mentioned allover including in IFC documentation ... but if the Door is already  directed listes in the BoundedBy set of IfcRelSpaceBoundaries i really don't see why going down that path of Opening/Voids/Fills in order to reach the Doors ... I'm still researching here which condition (in IFC export schema or parameters in the Design software... ) would lead to the situation where Doors / Windows are not listed directly in the Space BoundedBy set ... (and so, the other longer approach would be necessary...).

> True. It's rather straightforward. An efficient (bvh with only spaces, windows as point queries) is provided here https://community.osarch.org/discussion/comment/8902/#Comment_8902

Yes, i saw this post (very interesting !)  and already adapted the code in it to list doors and use BVH to get the spaces it provides access to ... it works perfectly ! ;-)

Thanks !
